File: examples/python/bond_custom_cashflows_example.py
#!/usr/bin/env python3
"""
Example demonstrating Bond instruments with custom cashflow schedules.

This example shows how to create bonds with complex cashflow patterns
using the cashflow builder and pass them to bond instruments.
"""

import finstack as fs
from finstack.instruments import Bond
from finstack.cashflow import (
    CashflowBuilder, 
    CouponPaymentType, 
    Amortization,
    cashflows_to_dataframe
)
import pandas as pd


def example_stepup_bond():
    """Create a bond with step-up coupons using custom cashflows."""
    print("\n=== Step-Up Bond with Custom Cashflows ===")
    
    # Create dates
    issue = fs.Date(2025, 1, 15)
    maturity = fs.Date(2028, 1, 15)
    
    # Build custom cashflow schedule
    # Note: Step-up rates would typically be implemented with multiple
    # fixed_coupon calls or custom windows (pending full API implementation)
    builder = CashflowBuilder()
    builder.principal(
        fs.Money(1_000_000, fs.Currency("USD")),
        issue,
        maturity
    )
    
    # For this example, using a single rate
    # In full implementation, would use builder.fixed_stepup() or similar
    builder.fixed_coupon(
        rate=0.04,  # Starting at 4%
        frequency=fs.Frequency.SemiAnnual,
        day_count=fs.DayCount.act365f(),
        payment_type=None,
        business_day_conv=None,
        calendar=None,
        stub=None
    )
    
    custom_schedule = builder.build()
    
    # Create bond from custom cashflows
    bond = Bond.from_cashflows(
        id="STEPUP_BOND_2028",
        schedule=custom_schedule,
        discount_curve="USD-OIS",
        quoted_clean_price=98.5
    )
    
    print(f"Created bond: {bond.id}")
    print(f"Issue date: {bond.issue_date}")
    print(f"Maturity: {bond.maturity}")
    print(f"Quoted clean price: {bond.quoted_clean_price}")
    
    # Analyze cashflows
    print(f"\nTotal cashflows: {len(custom_schedule.flows)}")
    df = cashflows_to_dataframe(custom_schedule)
    print("\nCashflow breakdown by type:")
    print(df.groupby('kind')['amount'].agg(['count', 'sum']))


def example_pik_toggle_bond():
    """Create a bond with PIK toggle features."""
    print("\n=== PIK Toggle Bond ===")
    
    issue = fs.Date(2025, 3, 1)
    maturity = fs.Date(2027, 3, 1)
    
    # Build schedule with PIK toggle
    builder = CashflowBuilder()
    builder.principal(
        fs.Money(10_000_000, fs.Currency("USD")),
        issue,
        maturity
    )
    
    # Create split payment type: 50% cash, 50% PIK
    split_type = CouponPaymentType.split(cash_pct=0.5, pik_pct=0.5)
    
    builder.fixed_coupon(
        rate=0.08,
        frequency=fs.Frequency.Quarterly,
        day_count=fs.DayCount.thirty360(),
        payment_type=split_type,
        business_day_conv=None,
        calendar=None,
        stub=None
    )
    
    custom_schedule = builder.build()
    
    # Create bond with custom cashflows in constructor
    bond = Bond(
        id="PIK_TOGGLE_2027",
        notional=fs.Money(10_000_000, fs.Currency("USD")),
        coupon=0.08,
        frequency=fs.Frequency.Quarterly,
        day_count=fs.DayCount.thirty360(),
        issue_date=issue,
        maturity=maturity,
        discount_curve="USD-OIS",
        custom_cashflows=custom_schedule  # Pass custom cashflows
    )
    
    print(f"Created PIK toggle bond: {bond.id}")
    print(f"Notional: {bond.notional}")
    print(f"Coupon rate: {bond.coupon:.2%}")
    
    # Analyze PIK vs cash flows
    pik_flows = custom_schedule.pik_flows()
    cash_flows = custom_schedule.coupons()
    
    print(f"\nCashflow breakdown:")
    print(f"  PIK flows: {len(pik_flows)}")
    if pik_flows:
        total_pik = sum(cf.amount for cf in pik_flows)
        print(f"  Total PIK amount: ${total_pik:,.2f}")
    
    print(f"  Cash coupon flows: {len(cash_flows)}")
    if cash_flows:
        total_cash = sum(cf.amount for cf in cash_flows)
        print(f"  Total cash coupons: ${total_cash:,.2f}")


def example_amortizing_bond():
    """Create an amortizing bond."""
    print("\n=== Amortizing Bond ===")
    
    issue = fs.Date(2025, 6, 1)
    maturity = fs.Date(2030, 6, 1)
    
    # Build cashflow schedule with amortization
    builder = CashflowBuilder()
    builder.principal(
        fs.Money(50_000_000, fs.Currency("EUR")),
        issue,
        maturity
    )
    
    # Linear amortization to 20% of original
    builder.with_amortization(
        Amortization.linear_to(
            fs.Money(10_000_000, fs.Currency("EUR"))
        )
    )
    
    # Fixed coupon
    builder.fixed_coupon(
        rate=0.045,
        frequency=fs.Frequency.SemiAnnual,
        day_count=fs.DayCount.act360(),
        payment_type=None,
        business_day_conv=fs.BusDayConvention.ModifiedFollowing,
        calendar="EUR",
        stub=None
    )
    
    custom_schedule = builder.build()
    
    # Create bond from cashflows
    bond = Bond.from_cashflows(
        id="AMORT_BOND_2030",
        schedule=custom_schedule,
        discount_curve="EUR-OIS",
        quoted_clean_price=None
    )
    
    print(f"Created amortizing bond: {bond.id}")
    print(f"Initial notional: {custom_schedule.notional}")
    
    # Show amortization profile
    principal_flows = custom_schedule.principal_flows()
    amort_flows = [cf for cf in principal_flows if cf.kind == "Amortization"]
    
    print(f"\nAmortization profile:")
    print(f"  Number of amortizations: {len(amort_flows)}")
    if amort_flows:
        total_amort = sum(abs(cf.amount) for cf in amort_flows)
        print(f"  Total amortization: EUR {total_amort:,.2f}")
    
    # Show outstanding path
    print("\nOutstanding principal over time (first 5):")
    for i, (date, amount) in enumerate(custom_schedule.outstanding_path()[:5], 1):
        print(f"  {i}. {date}: EUR {amount}")


def example_comparison_regular_vs_custom():
    """Compare regular bond vs bond with custom cashflows."""
    print("\n=== Regular Bond vs Custom Cashflow Bond ===")
    
    issue = fs.Date(2025, 1, 1)
    maturity = fs.Date(2026, 1, 1)
    
    # Create regular bond
    regular_bond = Bond(
        id="REGULAR_BOND",
        notional=fs.Money(1_000_000, fs.Currency("USD")),
        coupon=0.05,
        frequency=fs.Frequency.Annual,
        day_count=fs.DayCount.act365f(),
        issue_date=issue,
        maturity=maturity,
        discount_curve="USD-OIS"
    )
    
    # Create custom cashflow schedule with higher frequency
    builder = CashflowBuilder()
    builder.principal(
        fs.Money(1_000_000, fs.Currency("USD")),
        issue,
        maturity
    )
    builder.fixed_coupon(
        rate=0.05,
        frequency=fs.Frequency.SemiAnnual,  # Higher frequency
        day_count=fs.DayCount.act365f(),
        payment_type=None,
        business_day_conv=None,
        calendar=None,
        stub=None
    )
    custom_schedule = builder.build()
    
    # Apply custom cashflows to create new bond
    custom_bond = regular_bond.with_cashflows(custom_schedule)
    
    print(f"Regular bond: {regular_bond.id}")
    print(f"  Frequency: {regular_bond.frequency}")
    print(f"  Coupon: {regular_bond.coupon:.2%}")
    
    print(f"\nBond with custom cashflows: {custom_bond.id}")
    print(f"  Original frequency: {custom_bond.frequency} (overridden)")
    print(f"  Custom schedule flows: {len(custom_schedule.flows)}")
    
    # Compare cashflow structures
    df = cashflows_to_dataframe(custom_schedule)
    print("\nCustom cashflow summary:")
    print(df[['date', 'amount', 'currency', 'kind']].head())


def example_dataframe_analysis():
    """Show how to analyze bond cashflows using DataFrames."""
    print("\n=== Cashflow DataFrame Analysis ===")
    
    issue = fs.Date(2025, 1, 1)
    maturity = fs.Date(2027, 1, 1)
    
    # Build a cashflow schedule
    builder = CashflowBuilder()
    builder.principal(
        fs.Money(5_000_000, fs.Currency("USD")),
        issue,
        maturity
    )
    builder.fixed_coupon(
        rate=0.06,
        frequency=fs.Frequency.Quarterly,
        day_count=fs.DayCount.act360(),
        payment_type=None,
        business_day_conv=None,
        calendar=None,
        stub=None
    )
    
    schedule = builder.build()
    
    # Create bond from schedule
    bond = Bond.from_cashflows(
        id="ANALYSIS_BOND",
        schedule=schedule,
        discount_curve="USD-OIS",
        quoted_clean_price=99.5
    )
    
    # Convert to DataFrame for analysis
    df = cashflows_to_dataframe(schedule)
    
    print("Cashflow DataFrame (first 5 rows):")
    print(df[['date', 'amount', 'currency', 'kind', 'accrual_factor']].head())
    
    # Summary statistics
    print("\nCashflow Summary:")
    print(f"  Total rows: {len(df)}")
    print(f"  Date range: {df['date'].min()} to {df['date'].max()}")
    print(f"  Total amount: ${df['amount'].sum():,.2f}")
    
    # Group by kind
    print("\nCashflows by type:")
    summary = df.groupby('kind')['amount'].agg(['count', 'sum', 'mean'])
    print(summary)
    
    # Timeline analysis
    print("\nQuarterly cashflow timeline:")
    df['year'] = pd.DatetimeIndex(df['date']).year
    df['quarter'] = pd.DatetimeIndex(df['date']).quarter
    quarterly = df.groupby(['year', 'quarter'])['amount'].sum()
    print(quarterly.head(8))


if __name__ == "__main__":
    print("=" * 60)
    print("Bond Instruments with Custom Cashflow Schedules")
    print("=" * 60)
    
    try:
        # Run examples
        example_stepup_bond()
        example_pik_toggle_bond()
        example_amortizing_bond()
        example_comparison_regular_vs_custom()
        example_dataframe_analysis()
        
        print("\n" + "=" * 60)
        print("Examples completed successfully!")
        print("=" * 60)
        print("\nNote: Full pricing with market context is pending")
        print("complete Python bindings implementation.")
        
    except Exception as e:
        print(f"\nError running examples: {e}")
        import traceback
        traceback.print_exc()

File: examples/rust/bond_custom_cashflows_example.rs
//! Example demonstrating Bond instruments with custom cashflow schedules.
//!
//! This example shows how to create bonds with complex cashflow patterns
//! using the cashflow builder and pass them to bond instruments for pricing.

use finstack_core::currency::Currency;
use finstack_core::dates::{BusinessDayConvention, Date, DayCount, Frequency, StubKind};
use finstack_core::market_data::multicurve::CurveSet;
use finstack_core::market_data::term_structures::discount_curve::DiscountCurve;
use finstack_core::money::Money;

use finstack_valuations::cashflow::amortization_notional::AmortizationSpec;
use finstack_valuations::cashflow::builder::{cf, CouponType, FixedCouponSpec, ScheduleParams};
use finstack_valuations::instruments::fixed_income::bond::Bond;
use finstack_valuations::traits::{CashflowProvider, Priceable};

use time::Month;

fn example_stepup_bond() -> finstack_core::Result<()> {
    println!("\n=== Step-Up Bond with Custom Cashflows ===");

    // Create dates
    let issue = Date::from_calendar_date(2025, Month::January, 15).unwrap();
    let maturity = Date::from_calendar_date(2028, Month::January, 15).unwrap();
    let step1 = Date::from_calendar_date(2026, Month::January, 15).unwrap();
    let step2 = Date::from_calendar_date(2027, Month::January, 15).unwrap();

    // Build custom cashflow schedule with step-up rates
    let schedule_params = ScheduleParams {
        freq: Frequency::semi_annual(),
        dc: DayCount::Act365F,
        bdc: BusinessDayConvention::Following,
        calendar_id: None,
        stub: StubKind::None,
    };

    let custom_schedule = cf()
        .principal(Money::new(1_000_000.0, Currency::USD), issue, maturity)
        .fixed_stepup(
            &[(step1, 0.03), (step2, 0.04), (maturity, 0.05)],
            schedule_params,
            CouponType::Cash,
        )
        .build()?;

    // Create bond from custom cashflows
    let bond = Bond::from_cashflows(
        "STEPUP_BOND_2028",
        custom_schedule.clone(),
        "USD-OIS",
        Some(98.5),
    )?;

    // Create market data
    let disc_curve = DiscountCurve::builder("USD-OIS")
        .base_date(issue)
        .knots([(0.0, 1.0), (1.0, 0.97), (2.0, 0.94), (3.0, 0.91)])
        .linear_df()
        .build()?;

    let curves = CurveSet::new().with_discount(disc_curve);

    // Price the bond
    let result = bond.price(&curves, issue)?;
    println!("Step-up bond PV: {}", result.value);
    if let Some(clean_price) = result.measures.get("clean_price") {
        println!("Clean price: {:.2}", clean_price);
    }
    if let Some(ytm) = result.measures.get("ytm") {
        println!("YTM: {:.4}%", ytm * 100.0);
    }

    // Show cashflow details
    let flows = bond.build_schedule(&curves, issue)?;
    println!("\nTotal cashflows: {}", flows.len());
    println!("First 5 flows:");
    for (i, (date, amount)) in flows.iter().enumerate().take(5) {
        println!("  Flow {}: {} -> {}", i + 1, date, amount);
    }
    if flows.len() > 5 {
        println!("  ... and {} more flows", flows.len() - 5);
    }

    Ok(())
}

fn example_pik_toggle_bond() -> finstack_core::Result<()> {
    println!("\n=== PIK Toggle Bond ===");

    let issue = Date::from_calendar_date(2025, Month::March, 1).unwrap();
    let maturity = Date::from_calendar_date(2027, Month::March, 1).unwrap();
    let toggle_date = Date::from_calendar_date(2026, Month::March, 1).unwrap();

    // Build schedule with PIK toggle
    let schedule_params = ScheduleParams {
        freq: Frequency::quarterly(),
        dc: DayCount::Thirty360,
        bdc: BusinessDayConvention::Following,
        calendar_id: Some("usd"),
        stub: StubKind::None,
    };

    let custom_schedule = cf()
        .principal(Money::new(10_000_000.0, Currency::USD), issue, maturity)
        .add_fixed_coupon_window(
            issue,
            toggle_date,
            0.08,
            schedule_params,
            CouponType::Split {
                cash_pct: 0.5,
                pik_pct: 0.5,
            },
        )
        .add_fixed_coupon_window(
            toggle_date,
            maturity,
            0.07,
            schedule_params,
            CouponType::Cash,
        )
        .build()?;

    // Create bond using builder pattern
    let bond = Bond::builder()
        .id("PIK_TOGGLE_2027")
        .cashflows(custom_schedule.clone())
        .disc_curve("USD-OIS")
        .build()?;

    // Create market data
    let disc_curve = DiscountCurve::builder("USD-OIS")
        .base_date(issue)
        .knots([(0.0, 1.0), (1.0, 0.96), (2.0, 0.92)])
        .linear_df()
        .build()?;

    let curves = CurveSet::new().with_discount(disc_curve);

    // Price the bond
    let result = bond.price(&curves, issue)?;
    println!("PIK toggle bond PV: {}", result.value);

    // Show outstanding path (PIK increases principal)
    let outstanding_path = custom_schedule.outstanding_path();
    println!("\nOutstanding principal path:");
    for (date, amount) in outstanding_path.iter().take(5) {
        println!("  {}: {}", date, amount);
    }
    if outstanding_path.len() > 5 {
        println!("  ... and {} more", outstanding_path.len() - 5);
    }

    Ok(())
}

fn example_amortizing_bond_with_fees() -> finstack_core::Result<()> {
    println!("\n=== Amortizing Bond with Fees ===");

    let issue = Date::from_calendar_date(2025, Month::June, 1).unwrap();
    let maturity = Date::from_calendar_date(2030, Month::June, 1).unwrap();

    // Build complex cashflow schedule
    let custom_schedule = cf()
        .principal(Money::new(50_000_000.0, Currency::EUR), issue, maturity)
        .amortization(AmortizationSpec::LinearTo {
            final_notional: Money::new(10_000_000.0, Currency::EUR),
        })
        .fixed_cf(FixedCouponSpec {
            coupon_type: CouponType::Cash,
            rate: 0.045,
            freq: Frequency::semi_annual(),
            dc: DayCount::Act360,
            bdc: BusinessDayConvention::ModifiedFollowing,
            calendar_id: Some("eur"),
            stub: StubKind::ShortFront,
        })
        .build()?;

    // Create bond from cashflows
    let bond = Bond::from_cashflows("AMORT_BOND_2030", custom_schedule.clone(), "EUR-OIS", None)?;

    // Create market data with extended knots for 5-year bond
    let disc_curve = DiscountCurve::builder("EUR-OIS")
        .base_date(issue)
        .knots([
            (0.0, 1.0),
            (1.0, 0.98),
            (2.0, 0.95),
            (3.0, 0.93),
            (4.0, 0.90),
            (5.0, 0.87),
            (6.0, 0.84),
        ])
        .linear_df()
        .build()?;

    let curves = CurveSet::new().with_discount(disc_curve);

    // Price the bond
    let result = bond.price(&curves, issue)?;
    println!("Amortizing bond PV: {}", result.value);

    // Show flow breakdown by type
    use finstack_valuations::cashflow::primitives::CFKind;
    use std::collections::HashMap;

    let mut flows_by_kind: HashMap<CFKind, (usize, f64)> = HashMap::new();
    for cf in &custom_schedule.flows {
        let (count, total) = flows_by_kind.get(&cf.kind).unwrap_or(&(0, 0.0));
        flows_by_kind.insert(cf.kind, (count + 1, total + cf.amount.amount()));
    }

    println!("\nCashflow breakdown by type:");
    for (kind, (count, total)) in flows_by_kind {
        println!("  {:?}: {} flows, total {:.2} EUR", kind, count, total);
    }

    Ok(())
}

fn example_comparison_regular_vs_custom() -> finstack_core::Result<()> {
    println!("\n=== Regular Bond vs Custom Cashflow Bond Comparison ===");

    let issue = Date::from_calendar_date(2025, Month::January, 1).unwrap();
    let maturity = Date::from_calendar_date(2026, Month::January, 1).unwrap();

    // Create regular bond
    let regular_bond = Bond {
        id: "REGULAR_BOND".to_string(),
        notional: Money::new(1_000_000.0, Currency::USD),
        coupon: 0.05,
        freq: Frequency::annual(),
        dc: DayCount::Act365F,
        issue,
        maturity,
        disc_id: "USD-OIS",
        quoted_clean: None,
        call_put: None,
        amortization: None,
        custom_cashflows: None,
        attributes: finstack_valuations::traits::Attributes::new(),
    };

    // Create custom bond with higher frequency
    let custom_schedule = cf()
        .principal(Money::new(1_000_000.0, Currency::USD), issue, maturity)
        .fixed_cf(FixedCouponSpec {
            coupon_type: CouponType::Cash,
            rate: 0.05,
            freq: Frequency::semi_annual(), // Higher frequency than regular bond
            dc: DayCount::Act365F,
            bdc: BusinessDayConvention::Following,
            calendar_id: None,
            stub: StubKind::None,
        })
        .build()?;

    let custom_bond = Bond::builder()
        .id("CUSTOM_BOND")
        .cashflows(custom_schedule)
        .disc_curve("USD-OIS")
        .build()?;

    // Create market data
    let disc_curve = DiscountCurve::builder("USD-OIS")
        .base_date(issue)
        .knots([(0.0, 1.0), (1.0, 0.98)])
        .linear_df()
        .build()?;

    let curves = CurveSet::new().with_discount(disc_curve);

    // Compare pricing
    let regular_result = regular_bond.price(&curves, issue)?;
    let custom_result = custom_bond.price(&curves, issue)?;

    println!("Regular bond (annual payments): {}", regular_result.value);
    println!(
        "Custom bond (semi-annual payments): {}",
        custom_result.value
    );

    let regular_flows = regular_bond.build_schedule(&curves, issue)?;
    let custom_flows = custom_bond.build_schedule(&curves, issue)?;

    println!("\nFlow comparison:");
    println!("  Regular bond flows: {}", regular_flows.len());
    println!("  Custom bond flows: {}", custom_flows.len());
    println!("  Difference: Custom bond has more frequent payments");

    Ok(())
}

fn main() -> finstack_core::Result<()> {
    println!("{}", "=".repeat(60));
    println!("Bond Instruments with Custom Cashflow Schedules");
    println!("{}", "=".repeat(60));

    example_stepup_bond()?;
    example_pik_toggle_bond()?;
    example_amortizing_bond_with_fees()?;
    example_comparison_regular_vs_custom()?;

    println!();
    println!("{}", "=".repeat(60));
    println!("Examples completed successfully!");
    println!("{}", "=".repeat(60));

    Ok(())
}
